### DISCUSSION TITLE
Make all gets batch gets borrowing the logic from https://github.com/graphql/dataloader

### DISCUSSION THREAD

**Discussion by onhate** [sensedeep/dynamodb-onetable#250] [Ideas]
https://github.com/graphql/dataloader is a very simple and helpful set of tricks that is capable of accumulating multiple async requests and run them in batch saving time and requests to the database.

For example, the followig if using dataloader logic would generate just a single batch get and response to each with it's specificly requested value.

```ts
await Promise([
   model.get({ pk: '1', sk: '1' }),
   model.get({ pk: '2', sk: '2' }),
   model.get({ pk: '2', sk: '2' }),
])
```

Or even...

```ts
async findByIds(ids: string[]) {
   return await Promise.all(id => model.get({ id }));
}
```

Now think about the benefits of it if your system supports multiple parallel requests at the same time, or if you have multiple promises fetching items by id in parallel.

**Comment by mobsense** [comment]
Have you tried:

```
let batch = {}
let [a, b, c] = await Promise(
    await model.get({...}, {batch}),
    await model.get({...}, {batch}),
    await model.get({...}, {batch}),
)

**Reply by onhate** [reply]
but isn't `model.batchGet(batch)` that returns the values?

the batch gets like you described works only if all get operations are on the same code execution, other than that I need to pass the batch parameter arround to as many levels of code it would need perform get operations, but imagine you could have something like the following and let the library manage all get operations that are running in parallel and group them in a single batch get operation, that would add much value.

```
await Promise.all([
   runSomeOperation(),
   runAnotherOperation()
])

```

**Reply by onhate** [reply]
I can work on a PoC for this

**Reply by mobsense** [reply]
Interested to see what you come up with.

**Comment by mobsense** [comment]
Sorry you are right:

```
let batch = {}
await model.get({...}, {batch})
await model.get({...}, {batch})
await model.get({...}, {batch})
let [a, b, c] = await Promise(await db.batchGet(batch))
```
